src/aead/chacha20_poly1305.rs

```diff
@@ -278,6 +279,89 @@ mod tests {
                     (CRYPTO_POLY1305_STATE_LEN + 255) / 256);
     }
 
+    // This verifies the encryption functionality provided by ChaCha20_ctr32
+    // is successful when either computed on distinct input/output buffers,
+    // or in-place.
+    //
+    // On certain architectures (x86 and some branches of ARM), the in-place
+    // operation only succeeds when the input/output buffers are exactly
+    // overlapping due to the assembly code. Hence we do not test offsets for 
+    // x86 and ARM architectures.
+    //
+    // See the body of |open| for more information and the workaround.
+    #[test]
+    pub fn chacha20_tests() {
+        test::from_file("src/chacha_tests.txt", |section, test_case| {
+            assert_eq!(section, "");
+
+            let key_bytes = test_case.consume_bytes("Key");
+            let mut key = [0u32; CHACHA20_KEY_LEN / 4];
+            for ki in 0..(CHACHA20_KEY_LEN / 4) {
+                let kb =
+                    slice_as_array_ref!(&key_bytes[ki * 4..][..4], 4).unwrap();
+
+                key[ki] = polyfill::slice::u32_from_be_u8(kb);
+            }
+
+            let ctr_bytes = test_case.consume_bytes("Ctr");
+            let mut ctr = [0u32; 4];
+            for ci in 0..4 {
+                let cb =
+                    slice_as_array_ref!(&ctr_bytes[ci * 4..][..4], 4).unwrap();
+
+                ctr[ci] = polyfill::slice::u32_from_be_u8(cb);
+            }
+
+            let input = test_case.consume_bytes("Input");
+            let output = test_case.consume_bytes("Output");
+
+            // Run the test case over all prefixes of the input
```

Comment:
s/input/input because the behavior of ChaCha20 implementation changes dependent on the length of the input./